test/js/src/main/scala/zio/test/mock/TestEnvironment.scala

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test.mock
+
+import scala.concurrent.ExecutionContext
+
+import zio._
+import zio.clock.Clock
+import zio.console.Console
+import zio.internal.PlatformLive
+import zio.random.Random
+import zio.scheduler.Scheduler
+import zio.system.System
+
+case class TestEnvironment(
+  clock: TestClock,
+  console: TestConsole,
+  random: TestRandom,
+  scheduler: TestScheduler,
+  system: TestSystem,
+) extends Clock
+    with Console
+    with Random
+    with Scheduler
+    with System
+
+object TestEnvironment {
+
+  val Value: Managed[Nothing, TestEnvironment] =
+    Managed.fromEffect {
+      for {
+        bootstrap <- ZIO.effectTotal(PlatformLive.fromExecutionContext(ExecutionContext.global))
+        clock     <- Ref.make(TestClock.DefaultData).map(TestClock(_))
+        console   <- Ref.make(TestConsole.DefaultData).map(TestConsole(_))
+        random    <- Ref.make(TestRandom.DefaultData).map(TestRandom(_))
+        scheduler = TestScheduler(clock.ref, Runtime(Clock(clock), bootstrap))
+        system    <- Ref.make(TestSystem.DefaultData).map(TestSystem(_))
+      } yield new TestEnvironment(clock, console, random, scheduler, system)
+    }
+
+  object testConsole {
+    def feedLines(lines: String*): ZIO[TestEnvironment, Nothing, Unit] =
```

Comment:
This is fine for this pull request, but in general, I think I'd put these helpers inside the companion objects of the `Test*` classes.

e.g.:

```scala
TestConsole.feedLines(...)
```

Then there's a logical place for users to look for them.

Reply:
Yes I think this makes sense. One thing I am thinking about is whether it makes sense to have something like:

```
trait TestConsole extends Console {
  val console: TestConsoleImplementation
}

case class TestConsoleImplementation extends Console.Service[Any]
```

That way when we define these helper methods we can only depend on the specific environment type that these methods require and users could get the benefit of the method if they only depend on certain mock objects instead of the mock environment. Not sure about naming. What do you think?

Reply:
Yes, you are right. Let's follow the ZIO conventions though:

```scala
trait MockConsole extends Console {
  val console: MockConsole.Service[Any]
}
object MockConsole {
  trait Service[R] extends Console.Service[R] { def feedLines(...) }

  def feedLines(lines: String*): ZIO[MockConsole, Nothing, Unit] = ...
}
```

Reply:
Great. I'm on it!